nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -722,152 +806,322 @@ impl BundleCollection {
+                            collection
+                                .collect_sp_dump(&mgs_client, sp, dir)
+                                .await
                         }
-                    })
-                    .await;
-                if let Some(Err(err)) = prev_result {
-                    warn!(&self.log, "{FAILURE_MESSAGE}"; "err" => ?err);
-                }
-            }
-            while let Some(result) = set.join_next().await {
-                if let Err(err) = result {
-                    warn!(&self.log, "{FAILURE_MESSAGE}"; "err" => ?err);
-                }
-            }
+                        .boxed()
+                    }
+                }),
+            ));
         }
 
-        if let Some((host, sp)) = ereport_collection {
-            let (host, sp) = tokio::join!(host, sp);
-            const TASK_FAILURE_MSG: &str = "task failed";
-            let n_collected =
-                self.host_ereports_collected.load(Ordering::Acquire);
-            report.host_ereports = match host
-                .map_err(|e| anyhow::anyhow!("{TASK_FAILURE_MSG}: {e}"))
-                .and_then(|x| x)
-            {
-                Ok(_) => SupportBundleEreportStatus::Collected { n_collected },
-                Err(err) => {
-                    warn!(
-                        &self.log,
-                        "Support bundle: host ereport collection failed \
-                         ({n_collected} collected successfully)";
-                        "err" => ?err,
-                    );
-                    SupportBundleEreportStatus::Failed {
-                        n_collected,
-                        error: err.to_string(),
+        Ok(CollectionStepOutput::Spawn { extra_steps })
+    }
+
+    async fn collect_sp_dump(
+        &self,
+        mgs_client: &MgsClient,
+        sp: SpIdentifier,
+        dir: &Utf8Path,
+    ) -> anyhow::Result<CollectionStepOutput> {
+        save_sp_dumps(mgs_client, sp, dir)
+            .await
+            .with_context(|| format!("SP {} {}", sp.type_, sp.slot))?;
+
+        Ok(CollectionStepOutput::SavingSpDumps { listed_sps: true })
+    }
+
+    // Perform the work of collecting the support bundle into a temporary directory
+    //
+    // - "dir" is a directory where data can be stored.
+    // - "bundle" is metadata about the bundle being collected.
+    //
+    // If a partial bundle can be collected, it should be returned as
+    // an Ok(SupportBundleCollectionReport). Any failures from this function
+    // will prevent the support bundle from being collected altogether.
+    //
+    // NOTE: The background task infrastructure will periodically check to see
+    // if the bundle has been cancelled by a user while it is being collected.
+    // If that happens, this function will be CANCELLED at an await point.
+    //
+    // As a result, it is important that this function be implemented as
+    // cancel-safe.
```

Comment:
We now get this property from the fact that anything spawned on the `ParallelTaskSet` is automatically aborted when `run_collect_bundle_steps` is cancelled, FWIW. It might be worth saying that explicitly someplace --- any code in a `step` will be forcefully aborted when that future is cancelled, provided that it only spawns other tasks on the `ParallelTaskSet` rather than by using "normal spawn". We should maybe explictly note that so that people don't put random `tokio::spawn`s in a step...